Simplify Autorotation Mode Switching and Consolidate Autorotation State in RSC

MattKear opened the pull request:
This PR does the following:
- Removes the linking of autorotation flight mode and the interlock.
- Consolidates the autorotation state into its own class within the RSC.
- Moves the bailout logic out of the autorotation mode and into RSC making it common to both the mode and manual autorotations.

Reasons for this:
- The linking of the flight mode switching and interlock was creating lots of edge cases that were proving to difficult to prevent.  This is simpler and just like any other flight mode **(this is still SITL only)**.
- Removing this link between interlock and flight mode means, in the future, we can add partial power reduction in assisted autorotations via the flight mode which will make initial setup and testing more gradual and less intimidating.
- All bailout out handling for the manual autorotation has been tidied up and consolidated into the RSC.  The autorotation flight mode will be changed to use this bailout logic in the RSC to remove duplication.

This PR is WIP.  I still have the following to address:

- [x] Get bailouts working in autorotation flight mode.
- [x] Fixup the autotests to accept this new way of entering the autorotation flight mode.
- [x] Add autorotation flight mode exit handling e.g. don't allow exiting the flight mode to non-manual throttle modes unless the interlock is engaged and the RSC has flagged the bailout as complete.

IamPete1 commented on libraries/AC_Autorotation/AC_Autorotation.cpp:
```diff
@@ -81,39 +79,30 @@ const AP_Param::GroupInfo AC_Autorotation::var_info[] = {
     // @Increment: 0.1
     // @User: Advanced
-    AP_GROUPINFO("HS_SENSOR", 9, AC_Autorotation, _param_rpm_instance, 0),
+    AP_GROUPINFO("HS_SENSOR", 8, AC_Autorotation, _param_rpm_instance, 0),
```
Changing the indexes here will result in params being loaded from eeprom wrong. Its fine to leave a gap, just leave a comment what the index 8 was used for so noone trys to use it again in the future.

MattKear replied:
This is SITL only at the moment though, so I figured it was ok as it is not actually on anyone's eeprom yet?

IamPete1 replied:
Ah, OK, as you say, breaking SITL for people is not such a big deal.

IamPete1 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -389,38 +402,19 @@ void AP_MotorsHeli_RSC::output(RotorControlState state)
-    } else {
-        ramp_time = _ramp_time;
-    }
+    float ramp_time = MAX(float(_ramp_time.get()), 1.0);
```
const everything you can.

IamPete1 reviewed (commented):
just a quick look.

MattKear commented:
Added to DevCallEU for a quick look over/general feedback.  Not looking to get this merged today.

peterbarker commented on ArduCopter/heli.cpp:
```diff
@@ -189,29 +189,25 @@ void Copter::heli_update_rotor_speed_targets()
 {
-    // check if flying and interlock disengaged
-    if (!ap.land_complete && !motors->get_interlock()) {
+    bool in_autorotaion_mode = false;
```
```suggestion
    bool in_autorotation_mode = false;
```

MattKear commented:
I am pretty happy with this PR now.

It has been tested IRL now.  For complete transparency, this code was actually tested with the code in #28209 on top.

The key bit that we need to ensure is working well, is the autorotation state inside the RSC as that code will be available to users to download on 4.6-dev once this is merged.  This plot illustrates the RSC autorotation state transitioning from "active" to "bailing out" to "stopped"  During test the vehicle bailed out from the glide smoothly. 
![image](https://github.com/user-attachments/assets/1eada744-008b-4cd2-b3f7-a0550f3dd272)

The setup that I tested in real life is using a governor in the ESC, and the autorotation signal window was used to invoke a rapid spool up.
![image](https://github.com/user-attachments/assets/2caf6189-4bdb-4762-8e13-37bd47d6c946)

I have tested this code with Heli's head speed governor, a lot in Real Flight.  It does work, however, I will say that Heli's governor does not play that nice with autorotations.  However, this is an issue with way that the governor is currently written.  The over-speed/under-speed faults are significantly more likely to be tripped when recovering from an autorotation and the governor just gives up at that point.  This is an existing issue, that is not a problem with this code being added.  I am simply raising this for awareness now, and I will address the issue with the governor in a future PR.

Also for complete transparency, since testing this IRL I have done some restructuring to move the H_RSC_AROT params to the RSC_Autorotation object.  I did not change the logic of the mode, and have re-tested in the sim.  It was just something I noticed in testing, that we should have the RSC_AROT params hidden if not enabled.

bnsgeyer commented:
@MattKear I agree with the premise of this PR.  It seems a little odd having to switch over to autorotate mode and then disable interlock to initiate the autorotation.  I have tested this in Real Flight sim.  I plan to look at the follow on PR this week.

bnsgeyer reviewed (approved):
@MattKear I looked through this a little more thoroughly.  It is a pretty extensive change, for sure.  A little surprised to see the new RSC_autorotation library but it looks like  it simplified a bunch in RSC and the motors library.

Did you check the cooldown feature to ensure it still works as designed.  I know there is a test for the turbine start but I'm not sure if there is one for the cooldown timer.

I am happy with this and approved but want to ensure the cooldown timer still works.

MattKear commented:
Thanks for taking a look at this.

Yeah, the rework was for a few reasons:
- To simplify the autorotation state within the RSC
- To handle the bailout for more RSC configs (we previously didn't handle external governors with bailout throttle very well) which is why.
- To rename the params so that they appear to users to be more broadly applicable to the autorotation in general not just manual autorotation. This has meant that the duplication of bailout handling in the flight mode and the RSC has been now been removed.

MattKear commented:
No, I have not tested the turbine cool down functionality.  What should I look for to see that it still works?

bnsgeyer commented:
@MattKear checking it is pretty easy. It is used for ICE or turbine engines.  You will need to set a non zero H_RSC_IDLE. It will raise that by 50% for the cooldown time specified in H_RSC_CLDWN in seconds after motor interlock is disabled (RSC State is Idle). You should also verify that it doesn’t use the cooldown timer feature if the aircraft is in autorotation.

MattKear commented:
@bnsgeyer ok, thanks Bill. I will test and report back.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -305,46 +312,32 @@ void AP_MotorsHeli_RSC::output(RotorControlState state)
+
+        // check if we need to use autorotation idle throttle
+        if (autorotation.get_idle_throttle(_idle_throttle)) {
+            // if we are here then we are autorotating
```
@MattKear this guarantees the aircraft is in autorotation even if the throttle idle value is zero. I guess if you’re in our autorotation, it returns a true value for this call?

MattKear replied:
If you are in the autorotation mode the function autorotation.get_idle_throttle() populates the _idle_throttle variable with the value set by H_RSC_AROT_IDLE and returns true.  This is then assigned to the control output and returns early.  if we are not in the autorotation then _idle_throttle is not modified and the autorotation.get_idle_throttle() returns false giving access to the subsequent cases.
In the case of both turbine and piston, they would have to set H_RSC_AROT_IDLE to match H_RSC_IDLE to maintain the engine at idle during the autorotation.

bnsgeyer replied:
@MattKear I fixed the cooldown feature in this PR.  it required that I change the functionality of get_idle_throttle(). One because when AROT_IDLE was zero then it would use the cooldown idle.  and two, for some reason it would always use the cooldown throttle even when I used the autorotation state.  Look at what I did and let me know what  you think.
@Ferruccio1984 checkout and test the changes.  I think that works as it was meant to work.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli.cpp:
```diff
@@ -147,6 +148,10 @@ const AP_Param::GroupInfo AP_MotorsHeli::var_info[] = {
     AP_GROUPINFO("COL_LAND_MIN", 32, AP_MotorsHeli, _collective_land_min_deg, AP_MOTORS_HELI_COLLECTIVE_LAND_MIN),
 
+    // @Group: RSC_AROT_
+    // @Path: ../AP_Motors/AP_MotorsHeli_RSC.cpp
```
```suggestion
    // @Path: ../AC_Autorotation/RSC_Autorotation.cpp
```
This should fix the test that isn't passing CI

MattKear commented:
@bnsgeyer I have rebased this now that PR #28310 has gone in.

I have added more cases to autotest ManAutorotation() to check all of the conditions that you presented on Discord.

This is what the ManAutorotation test looks like:
![image](https://github.com/user-attachments/assets/e6d1c669-4e79-4609-b135-76318f4a16c9)

**Test 1**
H_RSC_IDLE = 5, H_RSC_AROT_RAMP = 2, H_RSC_AROT_IDLE = 0, H_RSC_CLDWN_TIME = 0
![image](https://github.com/user-attachments/assets/b4649a66-482f-4afe-959f-1348dc8023b2)

**Test 2**
H_RSC_IDLE = 0, H_RSC_AROT_RAMP = 0, H_RSC_AROT_IDLE = 20, H_RSC_CLDWN_TIME = 0
![image](https://github.com/user-attachments/assets/a3388bda-eb98-4f70-b0a5-092f05cf356b)

**Test 3**
H_RSC_IDLE = 5, H_RSC_AROT_RAMP = 2, H_RSC_AROT_IDLE = 0, H_RSC_CLDWN_TIME = 5
![image](https://github.com/user-attachments/assets/a52ee527-368a-460a-9fe3-db8e5ec1f03b)

**Test 4**
H_RSC_IDLE = 5, H_RSC_AROT_RAMP = 2, H_RSC_AROT_IDLE = 10, H_RSC_CLDWN_TIME = 5
![image](https://github.com/user-attachments/assets/aa215b71-71db-459a-bdef-0d5048f32af3)

MattKear commented:
@peterbarker I have also done the modifications that you recommended in PR #28310 to the turbine cooldown autotest.  This is kept as a separate commit.

bnsgeyer reviewed (approved):
LGTM

peterbarker commented on Tools/autotest/helicopter.py:
```diff
@@ -322,85 +326,169 @@ def AutoRotation(self, timeout=600):
+        self.progress("testing autorotation with cool down enabled and zero autorotation idle")
+        TestAutorotationConfig(self, rsc_idle=5.0, arot_ramp_time=2.0, arot_idle=0, cool_down=5.0)
+
+        self.progress("testing that H_RSC_AROT_IDLE is used over RSC_IDLE when cool down is enabled")
```
There's also `start_subtest` which adds a bit more delineation in the logs
